src/file_pipe_log/pipe_builder.rs

```diff
@@ -181,53 +193,125 @@ impl<F: FileSystem> DualPipesBuilder<F> {
+    }
+
+    /// Returns a mutable reference to the file_list related to `[LogQueue]`.
+    pub fn get_mut_file_list(&mut self, queue: LogQueue) -> &mut Vec<FileHandler<F>> {
+        if queue == LogQueue::Append {
+            &mut self.append_files
+        } else {
+            &mut self.rewrite_files
+        }
+    }
+}
+
+impl<F: FileSystem> DualPipesBuilder<F> {
+    /// Manually reads through log items in all available log files of the
+    /// specified queue, and replays them to specific [`ReplayMachine`]s
+    /// that can be constructed via `machine_factory`.
     pub fn recover_queue<M: ReplayMachine, FA: Factory<M>>(
-        &self,
-        queue: LogQueue,
+        recovery_cfg: RecoveryConfig<F>,
+        files: &mut Vec<FileHandler<F>>,
         replay_machine_factory: &FA,
         concurrency: usize,
     ) -> Result<M> {
-        let files = if queue == LogQueue::Append {
-            &self.append_files
-        } else {
-            &self.rewrite_files
-        };
         if concurrency == 0 || files.is_empty() {
             return Ok(replay_machine_factory.new_target());
         }
 
-        let recovery_mode = self.cfg.recovery_mode;
+        let recovery_mode = recovery_cfg.mode;
+        let recovery_read_block_size = recovery_cfg.recovery_read_block_size as usize;
+        let queue = recovery_cfg.queue;
+        let filesystem_ref = recovery_cfg.file_system.as_ref();
         let max_chunk_size = std::cmp::max((files.len() + concurrency - 1) / concurrency, 1);
-        let chunks = files.par_chunks(max_chunk_size);
+        let chunks = files.par_chunks_mut(max_chunk_size);
         let chunk_count = chunks.len();
         debug_assert!(chunk_count <= concurrency);
         let sequential_replay_machine = chunks
             .enumerate()
             .map(|(index, chunk)| {
                 let mut reader =
-                    LogItemBatchFileReader::new(self.cfg.recovery_read_block_size.0 as usize);
+                    LogItemBatchFileReader::new(recovery_read_block_size);
                 let mut sequential_replay_machine = replay_machine_factory.new_target();
                 let file_count = chunk.len();
-                for (i, f) in chunk.iter().enumerate() {
+                for (i, f) in chunk.iter_mut().enumerate() {
                     let is_last_file = index == chunk_count - 1 && i == file_count - 1;
                     if let Err(e) = reader.open(
                         FileId { queue, seq: f.seq },
-                        build_file_reader(self.file_system.as_ref(), f.handle.clone())?,
+                        // Attention please, to make sure that the `[Err]` message coud be captured,
+                        // the reader is just a `[LogFileReader]` without loadin the file header
+                        // in advance, by passing `Version::V1` as the default param.
+                        build_file_reader(filesystem_ref, f.handle.clone(), Some(Version::default()))?,
```

Comment:
Let me clarify a bit about the use of `build_file_reader`. We only pass in a non-null `version` field into the reader if we already know the format version of the file. Other time use `None` and let reader do the parsing itself (not doing it in `LogItemBatchFileReader`).

Reply:
Yes, u'r right, it would make the implementation confusing.
The reason why it inroduce a default `Version` was intended to let the `LogItemBatchFileReader.open(_)` do the `build` its `header` as the previous codes do. And this implementation would follow the error capture of `if let Err(e)`.
It could be modified with a better way.